expression/aggregation/base_func.go

```diff
@@ -357,17 +357,14 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
 		// function.
 		// Note: If the `Tp` of argument is the same as the `Tp` of the
 		// aggregation function, it will not wrap cast function on it
-		// internally. The reason of the special handling for `Column` is
-		// that the `RetType` of `Column` refers to the `infoschema`, so we
-		// need to set a new variable for it to avoid modifying the
-		// definition in `infoschema`.
-		if col, ok := a.Args[i].(*expression.Column); ok {
-			col.RetType = types.NewFieldType(col.RetType.Tp)
+		// internally.
+		switch x := a.Args[i].(type) {
+		case *expression.Column:
+			x.RetType = a.RetTp
+		case *expression.ScalarFunction:
+			x.RetType = a.RetTp
+		case *expression.CorrelatedColumn:
+			x.RetType = a.RetTp
```

Comment:
Do we need to copy the `RetTp` for safety?

Reply:
No, we can't 100% ensure it's not updated in other places. 
There are too many usages like here. If we use clone() every time, the performance would be worst.